Implement operator = and hashcode for CupertinoThemeData

jacobsimionato opened the pull request:
This fixes #71222

* Implement operator = and hashcode for CupertinoThemeData and CupertinoTextThemeData which CupertinoThemeData depends on.
* Add basic tests for CupertinoThemeData equality and hashcode.

I also manually tested the case mentioned in #71222, e.g. equality of ThemeData(cupertinoOverrideTheme: someTheme)

## Pre-launch Checklist

- [ x ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ x ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ x ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ x ] I signed the [CLA].
- [ x ] I listed at least one issue that this PR fixes in the description above.
- [ x ] I updated/added relevant documentation (doc comments with `///`).
- [ x ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ x ] All existing and new tests are passing.

@xster could you take a look at this? :-D

jacobsimionato committed:
Implement operator = and hashcode for CupertinoThemeData and CupertinoTextThemeData.

This fixes #71222

xster commented:
There's a linter that runs on the CI. 

Locally, please run `flutter analyze --flutter-repo` (or click into the Details link of the failing analyze test)

xster commented:
For the framework test 

```
00:25 +360: /b/s/w/ir/x/w/flutter/packages/flutter/test/material/theme_test.dart: Cupertino theme Cupertino theme override blocks derivative changes                                                   
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure object was thrown running a test:
  Expected: <2>
  Actual: <1>

When the exception was thrown, this was the stack:
#4      main.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/x/w/flutter/packages/flutter/test/material/theme_test.dart:594:9)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

This was caught by the test expectation on the following line:
  file:///b/s/w/ir/x/w/flutter/packages/flutter/test/material/theme_test.dart line 594
The test description was:
  Cupertino theme override blocks derivative changes
════════════════════════════════════════════════════════════════════════════════════════════════════
```

you can also run locally via `flutter test packages/flutter/test/material/theme_test.dart`

xster commented on packages/flutter/test/cupertino/theme_test.dart:
```diff
@@ -206,6 +206,41 @@ void main() {
   });
 
+  test('Theme operator == returns correct value for simple cases', () {
+    expect(const CupertinoThemeData(), equals(const CupertinoThemeData()));
```
nit (and unfortunate that our automatic formatter doesn't do it for you): but for consistent styling, please keep the line formatting consistent with the rest of the file. Mainly that closing parenths/brackets etc are in matching indents as the opening parenths/brackets. You can trigger it in the formatter by using trailing commas

xster reviewed (approved):
LGTM besides tests

jacobsimionato committed:
Fix some tests

jacobsimionato committed:
Fix always_put_control_body_on_new_line check in cupertino/theme.dart

This is in accordance with https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#separate-the-if-expression-from-its-statement

xster commented:
Looks like everything will pass :)

jacobsimionato commented:
It looks like Linux deferred components tests are flaky so I'll continue with the merge. https://ci.chromium.org/p/flutter/builders/prod/Linux%20deferred%20components
